### PR TITLE
Remove Eigen from basix.h

### LINKED ISSUE #96: Remove EIgen from public interface and use plain C API
Exposing Eigen in the public interface can cause all sort of problems when interfacing with other libraries that may use a different alignment. Simple fix is to not expose Eigen.

### PR DESCRIPTION
**Pull request by mscroggs**
Fixes #96.

### CODE REVIEW

File: cpp/basix.cpp [hunk 1]
```diff
@@ -117,22 +179,28 @@ const char* basix::mapping_name(int handle)
...
-std::vector<std::vector<std::vector<int>>>
-basix::topology(const char* cell_type)
+int basix::cell_geometry_dimension(const char* cell_type)
 {
   cell::type ct = cell::str_to_type(cell_type);
-  return cell::topology(ct);
+  return cell::geometry(ct).cols();
+}
+
+void basix::cell_geometry(const char* cell_type, double* points)
```
Comment (garth-wells) [inline cpp/basix.cpp:194]: Does this function get called frequently? (had a dynamic memory allocation in it)

File: cpp/finite-element.cpp [hunk 2]
```diff
@@ -159,6 +156,30 @@ const std::vector<std::vector<int>>& FiniteElement::entity_dofs() const
...
+  int ndsize = 1;
+  for (int i = 1; i <= nd; ++i)
+    ndsize *= (tdim + i);
+  for (int i = 1; i <= nd; ++i)
+    ndsize /= i;
+
+  const int ndofs = _coeffs.rows();
+  const int vs = value_size();
+
+  std::vector<double> basis_data(ndsize * x.rows() * ndofs * vs);
```
Comment (garth-wells) [inline cpp/finite-element.cpp:170]: Can we avoid this dynamic allocation?

File: cpp/finite-element.cpp [hunk 3]
```diff
@@ -170,44 +191,51 @@ FiniteElement::tabulate(int nd, const Eigen::ArrayXXd& x) const
...
   return _base_permutations;
 }
 //-----------------------------------------------------------------------------
+int FiniteElement::num_points() const { return _points.rows(); }
+//-----------------------------------------------------------------------------
 const Eigen::ArrayXXd& FiniteElement::points() const { return _points; }
 //-----------------------------------------------------------------------------
-Eigen::ArrayXd
-FiniteElement::map_push_forward(const Eigen::ArrayXd& reference_data,
+Eigen::ArrayXXd
```
Comment (garth-wells) [inline cpp/finite-element.cpp:217]: Should this be an in-out argument?

File: cpp/finite-element.cpp [hunk 4]
```diff
@@ -170,44 +191,51 @@ FiniteElement::tabulate(int nd, const Eigen::ArrayXXd& x) const
...
+    result.col(i) = mapping::map_push_forward(reference_data.col(i), J, detJ, K,
+                                              _mapping_type, _value_shape);
+  return result;
 }
 //-----------------------------------------------------------------------------
-Eigen::ArrayXd FiniteElement::map_pull_back(const Eigen::ArrayXd& physical_data,
-                                            const Eigen::MatrixXd& J,
-                                            double detJ,
-                                            const Eigen::MatrixXd& K) const
+Eigen::ArrayXXd
```
Comment (garth-wells) [inline cpp/finite-element.cpp:229]: Out argument?

File: cpp/finite-element.cpp (hunk 3 above)
Comment (mscroggs) [reply cpp/finite-element.cpp:217]: Yes, I've added this to #119

File: cpp/finite-element.cpp (hunk 4 above)
Comment (mscroggs) [reply cpp/finite-element.cpp:229]: Yes, I've added this to #119

File: cpp/finite-element.cpp (hunk 2 above)
Comment (mscroggs) [reply cpp/finite-element.cpp:170]: This version of tabulate is only used by the Python wrapper. The c-interface uses `tabulate_to_memory`

File: cpp/basix.cpp (hunk 1 above)
Comment (mscroggs) [reply cpp/basix.cpp:194]: No, this was changed to get rid of Eigen. I believe dolfinx will call this once then store the points

**Review by chrisrichardson — APPROVED** [review]

**Review by garth-wells — APPROVED** [review]

**Pull request merged**